### ISSUE TITLE
DHT queries

### ISSUE THREAD

**Issue by sim590** [savoirfairelinux/opendht#43]
Traffic usage has not been concern as of yet, but with data persistence functionality, it is a major issue that has to be addressed. In order to do so, here is the idea of filtering values by field values and field themselves. The following table represents a `Dht::Value`:

| Id  | ValueType | Data | OwnerPk | RecipientHash | UserType | Signature | EncryptedData |
| --- | --------- | ---- | ------- | ------------- | -------- | --------- | ------------- |
| ... | ...       | ...  | ...     | ...           | ...      | ...       | ...           |
| ... | ...       | ...  | ...     | ...           | ...      | ...       | ...           |
| ... | ...       | ...  | ...     | ...           | ...      | ...       | ...           |

Operations `Dht::get` and `Dht::put` could optionally ask for a parameter specifying filters over rows and columns of the above table. *However, this feature depends entirely on the completion of #42, since two separate queries over a same hash would have to be distinguished since they would not be necessarely satisfied in the same way.*

**Comment by sim590** [comment]
I have been working on this. [Here](https://github.com/sim590/opendht/commits/queries)'s what I have done so far. Packing, unpacking and remote handling of queries is done. What's missing is the part where queries are actually locally processed and sent to nodes. Handling of the response, i.e. associating the query's response to the callback and filter to apply, is also missing. The following is summarizing an abstract model of how to handle this.

![conception](https://github.com/sim590/opendht/raw/conception/doc/conception/queries/concept.png)

**NOTE**
- A `Query` distinguishes one `Get` structure instance from another. The same is true for `LocalListener` structure;
- Upon receiving an answer for a query, you only call callbacks for which the associated query is satisfied by the query to which the response is targeted;
- Request structure doesn't actually have a field `std::shared_ptr q`. The query shall only be captured into the `Request::done_cb` lambda.
- In order to track which `SearchNode` has already been sent a given query, we keep a pointer of query associated to a network engine request using a map:

  ```cpp
  std::map<std::shared_ptr<Query>, std::shared_ptr<Request>>
  ```
  That's why I use `std::shared_ptr<Query>`.

@aberaud: Do you have any thoughts? In the meantime, I'll be implementing this.

**Comment by sim590** [comment]
The code has reached a mature state. I'll be opening a PR shortly.

**Comment by sim590** [comment]
#70 has been merged, but now there are some enhancements to add to it. As mentionned in 52b4f5a, the `SELECT` feature is unusable in many cases. I'm working on enhancing queries by splitting Queries into two seperate independant structures which are `Select` and `Where`. Typical `get` call from the API will let the user pass a `Where` and there'll be an additionnal call `Dht::query` that will accept a Query as parameter. It's 'get' callback will have a diffrent signature to solve the issue stated [here](https://github.com/savoirfairelinux/opendht/pull/70#issuecomment-222507513).

**Comment by kaldoran** [comment]
I think this issue is resolved since the merge of PR #79 into master :sunglasses:
